wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+<?php
+
+require_once __DIR__ . '/class-mysql-token.php';
+
+/**
+ * This lexer is based on the MySQL Workbench lexer grammar.
+ * See:
+ *   https://github.com/mysql/mysql-workbench/blob/8.0.38/library/parsers/grammars/MySQLLexer.g4
+ *   https://github.com/mysql/mysql-workbench/blob/8.0.38/library/parsers/grammars/predefined.tokens
+ *   https://github.com/mysql/mysql-workbench/blob/8.0.38/library/parsers/mysql/MySQLBaseLexer.cpp
+ */
+class MySQL_Lexer {
+	// Token channels
+	const CHANNEL_DEFAULT = 0;
+	const CHANNEL_HIDDEN  = 99;
+
+	// SQL modes
+	const SQL_MODE_NO_MODE              = 0;
+	const SQL_MODE_ANSI_QUOTES          = 1 << 0;
+	const SQL_MODE_HIGH_NOT_PRECEDENCE  = 1 << 1;
+	const SQL_MODE_PIPES_AS_CONCAT      = 1 << 2;
+	const SQL_MODE_IGNORE_SPACE         = 1 << 3;
+	const SQL_MODE_NO_BACKSLASH_ESCAPES = 1 << 4;
+
+	// Whitespace characters
+	const WHITESPACES = array(
+		' '  => true,
+		"\t" => true,
+		"\n" => true,
+		"\r" => true,
+		"\f" => true,
+	);
+
+	/**
+	 * Unquoted identifiers:
+	 *   https://dev.mysql.com/doc/refman/8.4/en/identifiers.html
+	 *
+	 * Rules:
+	 *   1. Allowed characters are ASCII a-z, A-Z, 0-9, $, _ and Unicode \x{0080}-\x{ffff}.
+	 *   2. Unquoted identifiers may begin with a digit but may not consist solely of digits.
+	 */
+	const PATTERN_UNQUOTED_IDENTIFIER = '[a-zA-Z0-9_$\x{80}-\x{ffff}]*[a-zA-Z_$\x{80}-\x{ffff}][a-zA-Z0-9_$\x{80}-\x{ffff}]*';
```

Comment:
This is likely fine as tokenizing isn't a bottleneck, but I thought I'd still mention this. I've seen a substantial performance improvement in WP_XML_Processor after [replacing a similar `preg_match` with a long `if` clause checking the ranges of specific UTF-8 codepoints](https://github.com/WordPress/wordpress-playground/pull/1972/commits/adf6eabc546305213d7825d5085b44a126ffe3c7#r1826550245).